tensorboard/data/server/server.rs

```diff
@@ -52,28 +52,29 @@ impl TensorBoardDataProvider for DataProviderHandler {
         &self,
         _request: Request<data::ListPluginsRequest>,
     ) -> Result<Response<data::ListPluginsResponse>, Status> {
-        let mut res: data::ListPluginsResponse = Default::default();
-        let runs = self.commit.runs.read().unwrap();
-        let mut plugin_names_set: HashSet<String> = HashSet::new();
-        for run_data in runs.values() {
-            let run_data = run_data.read().unwrap();
-            for scalar_value in run_data.scalars.values() {
-                let metadata: &pb::SummaryMetadata = scalar_value.metadata.as_ref();
-                let plugin_data = &metadata.plugin_data.as_ref();
-                match plugin_data {
-                    Some(d) => {
-                        plugin_names_set.insert(d.plugin_name.clone());
-                    }
-                    _ => {
-                        plugin_names_set.insert(String::new());
-                    }
-                }
+        let runs = self.read_runs()?;
+        // Collect set of plugin names.
+        let mut plugin_names = HashSet::new();
+        for (run, data) in runs.iter() {
+            let data = data
+                .read()
+                .map_err(|_| Status::internal(format!("failed to read run data for {:?}", run)))?;
+            for time_series in data.scalars.values() {
+                let metadata: &pb::SummaryMetadata = time_series.metadata.as_ref();
+                let plugin_name = match &metadata.plugin_data.as_ref() {
```

Comment:
Can just match on `&metadata.plugin_data`. Then the `d` in `Some(d)` is
of type `&PluginData`, so `d.plugin_name.clone()` is still fine.

(Could also just match on `metadata.plugin_data.as_ref()` without the
extra indirection, which amounts to the same thing in this case but is
longer.)